src/pip/_internal/index.py

```diff
@@ -70,7 +102,68 @@ def _get_content_type(url, session):
     resp = session.head(url, allow_redirects=True)
     resp.raise_for_status()
 
-    return resp.headers.get("Content-Type", "")
+    _raise_for_non_html(resp.headers)
+
+
+def _get_html_response(url, filename, session):
+    """Access an HTML page with GET, and return the response.
+
+    This consists of five parts:
+
+    1. Check whether the URL looks suspective. If it does, make a HEAD
+       request to check the Content-Type header is indeed HTML, and raise
+       `_NotHTML` if it's not. Raise HTTP exceptions on network failures.
+    2. If URL scheme is file: and points to a directory, make it point to
+       index.html instead.
+    3. Actually perform the request. Raise HTTP exceptions on network failures.
+    4. Check whether Content-Type header to make sure the thing we got is HTML,
+       and raise `_NotHTML` if it's not.
+    """
+    if _is_filename_like_archive(filename):
+        _raise_for_non_html_response(url, session=session)
+
+    logger.debug('Getting page %s', url)
+
+    # Tack index.html onto file:// URLs that point to directories
+    scheme, _, path, _, _, _ = urllib_parse.urlparse(url)
+    if (scheme == 'file' and os.path.isdir(urllib_request.url2pathname(path))):
+        # add trailing slash if not present so urljoin doesn't trim
+        # final segment
+        if not url.endswith('/'):
+            url += '/'
+        url = urllib_parse.urljoin(url, 'index.html')
+        logger.debug(' file: URL is directory, getting %s', url)
```

Comment:
Btw, is the above block re-doing the `isdir()` check from before? If so, can the result of that check be stored along with the URL when the check is done the first time, or alternatively can `index.html` be appended when it's going through the first time? That way you wouldn't need to go back and forth with `url2pathname()`, etc.

Reply:
No, this applies when expand_dir is false, so is basically mutually exclusive with os.lsdir(). It is super confusing, and I want to sort it out after I’m done with HTMLPage.

Reply:
I agree this can / should be done later. I was just mentioning it because I noticed it. What I was referring to is that `isdir()` is [called in the code here](https://github.com/pypa/pip/blob/b6bbabe3e7bc633f7af9da35f55d8c9d0d7a3f09/src/pip/_internal/index.py#L328-L334) before `expand_dir` is checked. So I was saying that you could do something like pass around with the URL the information of whether or not it's a directory. That way you wouldn't have to convert the URL to a path again inside this function -- you just do it once. What I don't know though is whether every URL goes through the code path I linked to in my earlier sentence.